torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,202 @@
+import contextlib
+import functools
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Generator, List, Optional, Set, Tuple
+
+import torch
+
+
+__all__ = ["TracingConfig"]
+
+
+@dataclass
+class TracingConfig:
+    """
+    Configurations used in ``ParamExecOrderWrapPolicy`` for symbolic tracing of a model.
+    tracer: An instance of ``torch.fx.Tracer`` that will be used to perform symbolic
+        tracing. ``tracer`` is default to be ``torch.fx.Tracer()``, but can also be instance
+        of some child class of ``torch.fx.Tracer``. For example, one may want to use ``HFTracer``
+        for models in Transformers:
+        .. _Transformers: https://huggingface.co/docs/transformers/index
+
+    concrete_args: Concrete arguments that should not be treated as ``torch.fx.Proxy``
+        when tracing the forward function.
+        ``concrete_args`` allows one to partially specialize the forward function,
+        including removing control flow or data structures.
+        ``concrete_args`` is also the parameter used in ``tracer.trace()``:
+        https://pytorch.org/docs/stable/fx.html#torch.fx.Tracer.trace
+    """
+
+    tracer: torch.fx.Tracer = torch.fx.Tracer()
+    concrete_args: Optional[Dict[str, Any]] = None
+
+
+class _ExecutionInfo:
+    """
+    Contains the execution order information in the model forward pass.
+
+    Attributes:
+        current_module: record the module that is currently being traced.
+
+        module_forward_order: a list of modules, where the ordering is based on
+            when their forward function is called.
+
+        param_exec_order: a list of parameters ordered based on their execution order.
+
+        _traced_param_set: a set containing all parameters that have been traced.
+
+        module_execution_info_dict: a dict that maps each module to a list of
+            tuple containing a module and a list of named parameters.
+            For a given module, each tuple:
+            1. either contains this module and part of its ``named_parameters`` that will be executed together,
+            2. or contains one of its child modules and all of the child module's ``named_parameters``.
+            The list of tuple is ordered based on the parameter execution order.
+    """
+
+    def __init__(self, root_module: torch.nn.Module) -> None:
```

Comment:
Curious about implementation reasoning. IIUC, this is not a `@dataclass` because we need to do some operations to prepare the attributes. However, this is indeed used as a dataclass after construction. Any thoughts on how does this design compare to making this a dataclass and having a separate function that creates `_ExecutionInfo` from `nn.Module`?

Reply:
Thanks! I changed this into a `dataclass`, and add a function `_init_execution_info` that creates `_ExecutionInfo` from `nn.Module`.